src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -124,40 +151,103 @@ public VisitResult(TypeWithState resultType, TypeSymbolWithAnnotations lValueTyp
         private TypeWithState ResultType
         {
             get => _visitResult.RValueType;
-            set
-            {
-                SetResult(resultType: value, lvalueType: value.ToTypeSymbolWithAnnotations());
-            }
+        }
+
+        private void SetResultType(BoundExpression expression, TypeWithState type)
+        {
+
+            SetResult(expression, resultType: type, lvalueType: type.ToTypeSymbolWithAnnotations());
         }
 
         /// <summary>
         /// Force the inference of the LValueResultType from ResultType.
         /// </summary>
-        private void UseRvalueOnly()
+        private void UseRvalueOnly(BoundExpression expression)
         {
-            ResultType = ResultType;
+            SetResult(expression, ResultType, ResultType.ToTypeSymbolWithAnnotations(), isLvalue: false);
         }
 
         private TypeSymbolWithAnnotations LvalueResultType
         {
             get => _visitResult.LValueType;
-            set
-            {
-                SetResult(resultType: value.ToTypeWithState(), lvalueType: value);
-            }
+        }
+
+        private void SetLvalueResultType(BoundExpression expression, TypeSymbolWithAnnotations type)
+        {
+            SetResult(expression, resultType: type.ToTypeWithState(), lvalueType: type);
         }
 
         /// <summary>
         /// Force the inference of the ResultType from LValueResultType.
         /// </summary>
-        private void UseLvalueOnly()
+        private void UseLvalueOnly(BoundExpression expression)
         {
-            LvalueResultType = LvalueResultType;
+            SetResult(expression, LvalueResultType.ToTypeWithState(), LvalueResultType, isLvalue: true);
         }
 
-        private void SetResult(TypeWithState resultType, TypeSymbolWithAnnotations lvalueType)
+        private void SetInvalidResult() => SetResult(expression: null, _invalidType, _invalidType.ToTypeSymbolWithAnnotations(), updateAnalyzedNullability: false);
+
+        private void SetResult(BoundExpression expression, TypeWithState resultType, TypeSymbolWithAnnotations lvalueType, bool updateAnalyzedNullability = true, bool? isLvalue = null)
         {
             _visitResult = new VisitResult(resultType, lvalueType);
+            if (updateAnalyzedNullability)
+            {
+                SetAnalyzedNullability(expression, _visitResult, isLvalue);
+            }
+        }
+
+        private static bool ShouldMakeNotNullRvalue(BoundExpression node, bool wasReachable) => node.IsSuppressed || node.HasAnyErrors || !wasReachable;
+
+        /// <summary>
+        /// Sets the analyzed nullability of the expression to be the given result.
+        /// </summary>
+        private void SetAnalyzedNullability(BoundExpression expr, VisitResult result, bool? isLvalue = null)
+        {
```

Comment:
>{ [](start = 8, length = 1)

We should assert the type matches the expected type. (See `VisitExpressionWithoutStackGuard`.)
```
Debug.Assert(AreCloseEnough(result.RValueType.Type, expr.Type);
```

Reply:
See the disabled assertion below.

---
In reply to: [266096745](https://github.com/dotnet/roslyn/pull/34017#discussion_r266096745) [](ancestors = 266096745)

Reply:
Would `AreCloseEnough()` allow enabling the assertion now?

---
In reply to: [266121801](https://github.com/dotnet/roslyn/pull/34017#discussion_r266121801) [](ancestors = 266121801,266096745)

Reply:
Unfortunately not currently: how we handle conversions means that this api gets called with the wrong node.

---
In reply to: [266126506](https://github.com/dotnet/roslyn/pull/34017#discussion_r266126506) [](ancestors = 266126506,266121801,266096745)